# 异常

比如我们定义了一个10大小的数组，下标最多到9，如果我们代码中有a[10]=10，运行时就会出现**异常**。

以如下代码为例：
````````````java
public static void main(String[] args) {
		int[] a = new int[10];
		int idx;
		Scanner in = new Scanner(System.in);
		idx = in.nextInt();
		try {
			a[idx] = 10;
			System.out.println("hello");
		}catch(ArrayIndexOutOfBoundsException e) {
			System.out.println("Caught");
		}
	}
``````````````
try里面的是可能出现异常的代码，在try后面的catch里处理所有可能出现的异常，然后会继续往下走，而不会回头去做没做完的事情。

![异常][1]

![异常][2]

![异常][3]

`printStackTrace()`:可以一步步追溯异常发生的位置

![异常][4]


  [1]: https://github.com/LibraTang/Pics/blob/master/Java-Notes/%E5%BC%82%E5%B8%B81.png
  [2]: https://github.com/LibraTang/Pics/blob/master/Java-Notes/%E4%B8%80%E5%9C%BA.png
  [3]: https://github.com/LibraTang/Pics/blob/master/Java-Notes/%E5%BC%82%E5%B8%B83.png
  [4]: https://github.com/LibraTang/Pics/blob/master/Java-Notes/%E5%BC%82%E5%B8%B84.png